### ISSUE TITLE
Deprecate host.load() and host.create()

### ISSUE THREAD

**Issue by mottosso** [getavalon/core#195]
### Goal

Delegate all host-specific responsibility into plug-ins.

### Motivation

At the moment, plug-ins determine company- and host-specific behaviour of the pipeline, *in addition to the host interface*.

```python
from mindbender import api, maya

# Run the host-specific installer
api.install(maya)

# It registers these members to be used by the include tools, the Loader and Creator.
maya.create("myInstance", family="rig")
maya.load("myAsset")
```

The `create()` function then determines, based on the family, which of the Creator plug-ins to run. Similarly, `load()` determines which Loader plug-ins to use based on the families of the published asset.

What these two functions do is almost entirely generic, with the exception of `load()` also performing a post-process of encapsulating all loaded nodes into a `container`, an object used to manage loaded assets in the host (in the case of Maya, an `objectSet`).

### Implementation

Delegate the one outlier from a completely generic system in `load()` to the superclass of the Loader plug-in.

We can then start looking at making menus and scripts part of the plug-in architecture as well, for a completely loosely coupled system with entirely customisable behavior and functionality.

**Comment by BigRoy** [comment]
I'm working on this right now.

As I see it the `create`, `load`, `remove` and `update` methods become the responsibility of the Loader plug-ins. A loader would look like:

```python
# pseudocode (the arguments of the methods are just placeholders here)
class Loader(api.Loader):
    def load(self, representation):
        pass
    def update(self, version=-1):
        pass
    def remove(self, container):
        pass
```

As such a `host` itself would not implement specific loading logic but the "static" logic would move into the core api of Avalon, as such the api would expose:

```python
# Create an instance for publishing
api.create(name, asset, family, options, data)

# Load
api.load(Loader, representation, namespace, name, data)

# Remove a container
api.remove(container)

# Change version of a container
api.update(container, version=2)
```

---

Here's a quick copy and paste of how it could be refactored into the core `pipeline.py`:

```python
def create(name, asset, family, options=None, data=None):
    """Create a new instance

    Associate nodes with a subset and family. These nodes are later
    validated, according to their `family`, and integrated into the
    shared environment, relative their `subset`.

    Data relative each family, along with default data, are imprinted
    into the resulting objectSet. This data is later used by extractors
    and finally asset browsers to help identify the origin of the asset.

    Arguments:
        name (str): Name of subset
        asset (str): Name of asset
        family (str): Name of family
        options (dict, optional): Additional options from GUI
        data (dict, optional): Additional data from GUI

    Raises:
        NameError on `subset` already exists
        KeyError on invalid dynamic property
        RuntimeError on host error

    Returns:
        Name of instance

    """

    host = registered_host()

    plugins = list()
    for Plugin in discover(Creator):
        has_family = family == Plugin.family

        if not has_family:
            continue

        Plugin.log.info(
            "Creating '%s' with '%s'" % (name, Plugin.__name__)
        )

        try:
            plugin = Plugin(name, asset, options, data)

            with host.maintained_selection():
                instance = plugin.process()
        except Exception as e:
            log.warning(e)
            continue

        plugins.append(plugin)

    assert plugins, "No Creator plug-ins were run, this is a bug"
    return instance


def load(Loader, representation, namespace=None, name=None, data=None):

    assert representation is not None, "This is a bug"

    if isinstance(representation, (six.string_types, io.ObjectId)):
        representation = io.find_one(
            {"_id": io.ObjectId(str(representation))})

    version, subset, asset, project = io.parenthood(representation)

    assert all([representation, version, subset, asset, project]), (
        "This is a bug"
    )

    context = {
        "project": project,
        "asset": asset,
        "subset": subset,
        "version": version,
        "representation": representation,
    }

    # TODO: Ensure the Loader is valid for the representation
    loader = Loader()
    loader.load(context, namespace, name, data)


def remove(container):
    """Remove a container"""

    loader = container['Loader']
    for Plugin in discover(Loader):

        # TODO: Ensure the loader is valid
        if Plugin.__name__ == loader:
            plugin = Plugin()
            plugin.remove(container)


def update(container, to_version=-1):
    """Update a container"""

    # Compute the different version from 'representation'
    current_representation = io.find_one({
        "_id": io.ObjectId(container["representation"])
    })

    assert current_representation is not None, "This is a bug"

    version_, subset, asset, project = io.parenthood(
        current_representation)

    if to_version == -1:
        new_version = io.find_one({
            "type": "version",
            "parent": subset["_id"]
        }, sort=[("name", -1)])
    else:
        new_version = io.find_one({
            "type": "version",
            "parent": subset["_id"],
            "name": to_version,
        })

    new_representation = io.find_one({
        "type": "representation",
        "parent": new_version["_id"],
        "name": current_representation["name"]
    })

    assert new_version is not None, "This is a bug"

    template_publish = project["config"]["template"]["publish"]
    fname = template_publish.format(**{
        "root": registered_root(),
        "project": project["name"],
        "asset": asset["name"],
        "silo": asset["silo"],
        "subset": subset["name"],
        "version": new_version["name"],
        "representation": current_representation["name"],
    })

    # Find the Loader to process the update
    loader = container['Loader']
    for Plugin in discover(Loader):

        # TODO: Ensure the loader is valid
        if Plugin.__name__ == loader:
            plugin = Plugin()
            plugin.update(container, new_representation, fname)
```

**Comment by BigRoy** [comment]
> Note: In this example the `host` implements a `maintained_selection` method so that the API can ensure all applications will hold True to that logic, even when the loader does not implement it. But that would require a `maintained_selection` context manager to be implemented for each host. - _This can be seen in the `create` method._

**Comment by BigRoy** [comment]
Implemented with #291
